drivers/builtin/src/asn1write.c

```diff
@@ -434,4 +436,52 @@ mbedtls_asn1_named_data *mbedtls_asn1_store_named_data(
 
     return cur;
 }
+
+size_t mbedtls_get_asn1_integer_frame_length(size_t integer_length){
+    size_t header_length =2;
+    if(integer_length > 126) {
+        header_length = 3;
+    } else if(integer_length > 254) {
+        header_length = 4;
+    }
+    return integer_length + header_length;
+}
+
+/*
+ * to test
+ * buffer too small
+ * buffer to big
+ * different integer length to buffer size
+ * Oversize integer length
+ */
+int mbedtls_asn1_write_integer(unsigned char **p, unsigned char *start, const unsigned char *integer, size_t integer_length, size_t buffer_length) {
+
+    mbedtls_mpi mpi_to_encode; //declare a mpi for storing the input
+    int result_size;
+    int ret =-1;
+
+
+    if((*p == NULL) || (start == NULL) || (integer == NULL)){
+        return MBEDTLS_ERR_ASN1_INVALID_DATA; //TC1 NULL Pointer exceptions.
+    }
+
+    if(integer_length > 0xFFFFFFFF){
+        return MBEDTLS_ERR_ASN1_INVALID_LENGTH; //TC2 Integer length too large.
+    }
+
+    if(buffer_length < mbedtls_get_asn1_integer_frame_length(integer_length)){
+	return MBEDTLS_ERR_ASN1_BUF_TOO_SMALL; //TC3 Buffer too small
+    }
+
+    mbedtls_mpi_init(&mpi_to_encode);
+    memset(start, 0, buffer_length);
+
+    ret = mbedtls_mpi_read_binary(&mpi_to_encode, integer, integer_length); //read the input string into the mpi
+    if(ret !=0){
+        return ret;
+    }
+    result_size = mbedtls_asn1_write_mpi(p, start, &mpi_to_encode); //encode the mpi into asn1
```

Comment:
Although we can use the MPI function internally for now, I would prefer to make this the primary function, and change `write_mpi` to call it. That way the new function would benefit from the existing tests and it would directly be more future-looking, and it would do no heap allocation.